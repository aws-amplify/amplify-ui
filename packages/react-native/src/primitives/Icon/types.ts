import { ColorValue, ImageProps, ImageStyle } from 'react-native';

export type Size = 'xs' | 'small' | 'medium' | 'large' | 'xl';
export type IconSizes = Record<Size, number>;

export interface IconStyles {
  icon?: ImageStyle;
}

export interface IconProps extends ImageProps {
  /**
   * Whether the icon is animated
   */
  animated?: boolean;

  /**
   * Color to be applied to the icon
   */
  color?: ColorValue;

  /**
   * Icon size
   */
  size?: Size | number;
}
