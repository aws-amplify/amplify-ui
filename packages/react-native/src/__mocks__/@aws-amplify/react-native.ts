export default {
  loadUrlPolyfill: jest.fn(),
};
