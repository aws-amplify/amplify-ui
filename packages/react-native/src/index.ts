export {
  Authenticator,
  AuthenticatorProps,
  useAuthenticator,
  UseAuthenticator,
  withAuthenticator,
  WithAuthenticatorOptions,
} from './Authenticator';
export * from './InAppMessaging';
export {
  Theme,
  ThemeProvider,
  defaultDarkModeOverride,
  defaultTheme,
  useTheme,
} from './theme';
