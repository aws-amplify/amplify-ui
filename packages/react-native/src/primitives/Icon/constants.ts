import { IconSizes } from './types';

export const iconSizes: IconSizes = {
  xs: 16,
  small: 20,
  medium: 24,
  large: 32,
  xl: 48,
};
