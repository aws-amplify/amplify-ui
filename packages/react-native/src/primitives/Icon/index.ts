export { default as Icon } from './Icon';
export { IconProps, IconStyles, IconSizes } from './types';
export { iconSizes } from './constants';
