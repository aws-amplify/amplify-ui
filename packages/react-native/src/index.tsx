import React from 'react';
import { Text, View } from 'react-native';

export function HelloWorld(): JSX.Element {
  return (
    <View>
      <Text>Hello World!</Text>
    </View>
  );
}
