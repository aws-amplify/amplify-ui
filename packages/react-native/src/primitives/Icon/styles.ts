import { StyleSheet } from 'react-native';
import { ConsoleLogger as Logger } from '@aws-amplify/core';

import { StrictTheme } from '../../theme';
import { IconProps, IconStyles } from './types';
import { iconSizes } from './constants';

const logger = new Logger('Icon-logger');

const getIconSize = (size: IconProps['size']): number => {
  if (!size) {
    return iconSizes.medium;
  }

  if (typeof size === 'number') {
    return size;
  }

  if (!iconSizes[size]) {
    logger.warn(
      `"${size}" is not a valid icon size. Available values are: ${Object.keys(
        iconSizes
      )}`
    );
    return iconSizes.medium;
  }
  return iconSizes[size];
};

export const getThemedStyles = (
  theme: StrictTheme,
  color?: IconProps['color'],
  size?: IconProps['size']
): Required<IconStyles> => {
  const { components } = theme;
  const iconSize = getIconSize(size);

  return StyleSheet.create({
    icon: {
      height: iconSize,
      resizeMode: 'contain',
      tintColor: color,
      width: iconSize,
      ...components?.icon,
    },
  });
};
