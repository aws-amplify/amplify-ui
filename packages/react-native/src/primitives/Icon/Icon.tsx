import React, { useMemo } from 'react';
import { Animated, Image } from 'react-native';

import { useTheme } from '../../theme';
import { getThemedStyles } from './styles';
import { IconProps } from './types';

export default function Icon({
  accessibilityRole = 'image',
  animated,
  color,
  size,
  style,
  ...rest
}: IconProps): JSX.Element {
  const theme = useTheme();
  const themedStyle = useMemo(
    () => getThemedStyles(theme, color, size),
    [theme, color, size]
  );

  const imageProps: IconProps = {
    ...rest,
    accessibilityRole,
    style: [themedStyle.icon, style],
  };

  if (animated) {
    return <Animated.Image {...imageProps} />;
  }
  return <Image {...imageProps} />;
}
