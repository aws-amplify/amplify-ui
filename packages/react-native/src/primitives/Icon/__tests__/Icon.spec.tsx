import React from 'react';
import { render } from '@testing-library/react-native';
import Icon from '../Icon';

const MOCK_SOURCE = { uri: 'mock.png' };

describe('Icon', () => {
  it('renders as expected', () => {
    const { toJSON } = render(<Icon source={MOCK_SOURCE} />);
    expect(toJSON()).toMatchSnapshot();
  });

  it('renders as expected with custom styles', () => {
    const customStyle = { tintColor: 'red' };
    const { toJSON } = render(
      <Icon source={MOCK_SOURCE} style={customStyle} />
    );
    expect(toJSON()).toMatchSnapshot();
  });

  it('renders an animated icon', () => {
    const { toJSON } = render(<Icon source={MOCK_SOURCE} animated />);
    expect(toJSON()).toMatchSnapshot();
  });
});
