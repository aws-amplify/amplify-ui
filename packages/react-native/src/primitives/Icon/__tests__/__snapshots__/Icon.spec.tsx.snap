// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Icon renders an animated icon 1`] = `
<Image
  collapsable={false}
  source={
    Object {
      "uri": "mock.png",
    }
  }
  style={
    Object {
      "height": undefined,
      "resizeMode": "contain",
      "tintColor": undefined,
      "width": undefined,
    }
  }
/>
`;

exports[`Icon renders as expected 1`] = `
<Image
  source={
    Object {
      "uri": "mock.png",
    }
  }
  style={
    Array [
      Object {
        "height": undefined,
        "resizeMode": "contain",
        "tintColor": undefined,
        "width": undefined,
      },
      undefined,
    ]
  }
/>
`;

exports[`Icon renders as expected with custom styles 1`] = `
<Image
  source={
    Object {
      "uri": "mock.png",
    }
  }
  style={
    Array [
      Object {
        "height": undefined,
        "resizeMode": "contain",
        "tintColor": undefined,
        "width": undefined,
      },
      Object {
        "tintColor": "red",
      },
    ]
  }
/>
`;
